### DISCUSSION TITLE
I have some questions...

### DISCUSSION THREAD

**Discussion by andrewdwallo** [abivia/plaid#4] [Q&A] [answered]
If I was going to try to use Plaid with Laravel Models, Migration Tables, etc. How would I go about using your package so that I can receive requests and render responses with plaid? I also wanted to know... How would I go about using your package with Plaid Link where a user can login with something like Jetstream and click on a button to use link and have all of their data come in through the user interface in tables for Models..? I am also planning on using Filament Admin for this... Sorry for all the questions but I am having a hard time trying to do this with Plaid and I am having a lot of errors in doing so... Also, does your package not provide a config file to publish so that I can retrieve your vendor files for models and such in the root folder? Maybe I am just a noob or something but how do I actually go about using your package... I would really appreciate your help, thanks!

**Comment by abivia** [comment, marked as answer]
Hi @andrewdwallo, thanks for your question!
There are no models. This is simply an interface to the Plaid API. The only configuration is the environment settings described at https://github.com/abivia/plaid#usage

For an example on how to use the interface, check out tests/Feature/TransactionTest.php It shows the process of obtaining an access token and then performing a call to Plaid's transactions/get entry point. The result is an array as described in the Plaid documentation at https://plaid.com/docs/

How your application handles that information, including storing it in your Models is entirely up to you.

**Reply by andrewdwallo** [reply]
Okay thank you and I understand it is up to me but I was just asking because I don't know how I would do it with laravel models and migrations and I guess would I use the TransactionTest.php file as a reference for a Controller file in App/Http/Controllers for get requests? and then would I pass that to a file or what?

**Reply by andrewdwallo** [reply]
Also what kind of laravel package would you suggest with something like plaid? Is Sanctum okay and will it work or do I need to use Passport? Also should I use Breeze with NextJs or react instead of livewire and blade?

**Reply by abivia** [reply]
Well, as a simple example, let's say you wanted to store just a few fields from the transaction response: date, amount, and name.

You would create a migration (for the table some_models) that had an id, the date, amount, and name as fields. Then if you get a list of transactions, step through to extract those fields, put the data into the model and save:

```
$response = Plaid::transactions()->list(/* application dependent arguments here */)
foreach ($response['transactions'] as $trans) {
    $myModel = new SomeModel();
    $myModel->date = $trans['date'];
    $myModel->amount = $trans['amount'];
    $myMode->name = $trans['name'];
    $myModel->save()
}
````

Now the data is stored in your database and available for doing whatever your application requires.

This code would be in your controller, but the specifics are highly application dependent. The event that triggered this process could be anything. A cron job, a button on a form...

**Reply by abivia** [reply]
Your questions about Sanctum/Passport et. al. are out of scope. So much depends on your preferences, environment, and so on that I'm not the right person to ask.

**Reply by andrewdwallo** [reply]
Okay cool. So in the code it says "$myModel = new SomeModel();" are those the same model?
For Example:
$Transaction = new Transaction();
or
would it be lowercase with something like this:?
$transactions = new Transaction();

**Reply by andrewdwallo** [reply]
with $transactions representing the table and $Transaction representing the model?

**Reply by andrewdwallo** [reply]
I just know a lot of people who have wanted to integrate something like this with laravel and I know it would be useful to provide something to those who are wanting to do something like this...

**Reply by abivia** [reply]
That's just a variable name. You could write ```$item = new Transaction()``` and have the same result. The PSR convention for variable names is camelCase, so something like ```$myNewItem``` would be consistent with the standards and keep your code more readable. Mixed case identifiers beginning with an uppercase character are for classes. Most people use snake_case for database fields but that's not part of the PSR standards. I personally prefer camelCase there too.

**Reply by andrewdwallo** [reply]
So when you say that your package is just an interface to plaid api.., does that mean like it is a view file like a blade file or?...

**Reply by andrewdwallo** [reply]
Like how would I go about even using your package in a laravel project?

**Reply by andrewdwallo** [reply]
and if I was going to reference one of your files in a controller or model class would I just put for example:

use Abivia/Plade;

at the top of the php file?

**Reply by abivia** [reply]
It's still highly implementation dependent. It's far from ready to show yet, but we're working on an application that integrates Plaid, Abivia Ledger, and some other systems. Not everyone will want to use Ledger for this, and how the Plaid transactions map into Ledger would still be application dependent, so it's hard to provide anything that's useful beyond example code.

That said, when you have something like that ready, let me know. Perhaps we can link to it.

**Reply by andrewdwallo** [reply]
alright will do.. so does that mean I should use your package for my project or something else? Like should I reference your file at the top of my Controller and Model classes and use your schema or?

**Reply by andrewdwallo** [reply]
And btw the reason why I am using something like Plaid is because I am creating an ERP System for Laravel so it would definitely have a lot of things like Ledgers and Accounting, Ecommerce Modules, etc so if you ever wanted to collab I am so down...

**Reply by abivia** [reply]
> and if I was going to reference one of your files in a controller or model class would I just put for example:
> 
> use Abivia/Plade;
> 
> at the top of the php file?

```use Abivia\Plaid\Facades\Plaid;``` gets the Facade and lets you use things like ```Plaid::transactions(...);```. Otherwise ```use Abivia\Plaid\Api\...``` will get you specific classes where "..." is the name of the class, as in ```use Abivia\Plaid\Api\Transactions;```

**Reply by abivia** [reply]
> And btw the reason why I am using something like Plaid is because I am creating an ERP System for Laravel so it would definitely have a lot of things like Ledgers and Accounting, Ecommerce Modules, etc so if you ever wanted to collab I am so down...

Take a look at Abivia\Cogs, which is still somewhat in flux. I'm hoping to build a component based ERP, so that you can plug in a "stock" customer base, interface to an external CRM system via API (for example Salesforce), or code your own (say for an ecommerce customer base) and have them all interoperate by adhering to the same interfaces.

The problem I have with existing ERP systems is that the requirements change dramatically with the size of the enterprise and what it does for a living. As a hosting business, our products have a lot of unique characteristics. No manufacturing business needs a product base with those characteristics, just as we don't need inventory management. Still, we all need a ledger and we all have customers... it would be nice if someone could configure a lean ERP that met their specific requirements merely by specifying those components in composer, and making some connections wouldn't it?

**Reply by andrewdwallo** [reply]
Yes it most definitely would. It would be nice for if like lets say you had a Accounting module as a composer package with a .config file and ServiceProvider than I could just composer require yourpackage and then publish config/views and then whala I have frotend view files, migration files, model files and everything for the Accounting module... man that would be nice.

**Reply by andrewdwallo** [reply]
I looked at your package Abivia\Cogs... how would these interfaces be implemented into a Laravel project?

**Reply by andrewdwallo** [reply]
And what do you mean by "interface"?

**Reply by andrewdwallo** [reply]
And when you said "the problem with existing ERP systems is that the requirements change drastically.., etc." that is why I am basically trying to build a Multi-tenant ERP SAAS Software System with Laravel so you would only have to change the code base for each Database depending on the companies requirements.

**Reply by andrewdwallo** [reply]
https://github.com/andrewdwallo/erpsaas
Also here is the project that I have been working on for a little over a week. You can go to the discussions page and look at the ideas I have and want I am planning on building... Maybe it can line up with what you wanted to do and we could collab... just let me know what you think of the ideas I have, thanks!

**Reply by abivia** [reply]
> And what do you mean by "interface"?

I'm referring to PHP interfaces: https://www.php.net/manual/en/language.oop5.interfaces.php

Much of the Cogs interfaces are based on open standards from OASIS, which is indicative of how I'm trying to structure things. For example there is a European standard for describing product catalogues that I'd like to map as closely as possible to, in hopes of facilitating better API interoperability.

I will see if I can get this thing in shape for a private repo and share it with you. There is currently much code that's specific to our environment that I'd rather not disclose in a public repo just yet.

**Reply by abivia** [reply]
We have similar goals but I'm not sure if they align 100%. Right now our focus is on meeting some of our critical business requirements.

The good news is that there is an opportunity to collaborate since Ledger covers almost all of what you want in the core accounting module: a double entry ledger and journal, with multiple currencies, sub-journals, and a trial balance, mappings for tax reporting, audit trails, and a statement generator. Missing functionality still includes proper period closing and likely a few other small things I haven't discovered yet.

Nevertheless, Abivia\Ledger was designed exactly for projects like yours. So many developers cobble together code that has accounting functionality without understanding the most basic principles of accounting that I've tried to give them an easy to use base that plugs in and won't cause their CFO to have an aneurysm. I'd be happy to work with you to make it a fully functional core upon which you (and hopefully others) can build their own solutions from club memberships to full ERP solutions.

**Reply by andrewdwallo** [reply]
Sounds great to me, how could I contact you? Do you have discord or something similar?

**Reply by abivia** [reply]
I've mostly been using Slack but now that it's Salesforce I'm looking at migrating. Give me a little time to figure out how to move forward and drop info in the repo I just invited you to (you should get it by email).

**Reply by abivia** [reply]
> Sounds great to me, how could I contact you? Do you have discord or something similar?

I've set up a Mattermost cloud instance. If you get your email to me I'll send you an invite You can DM Abivia on Twitter or post in the private repo.
